driver-core/src/main/com/mongodb/internal/connection/SaslAuthenticator.java

```diff
@@ -131,6 +112,50 @@ private void throwIfSaslClientIsNull(final SaslClient saslClient) {
         }
     }
 
+    private BsonDocument getNextSaslResponse(final SaslClient saslClient, final InternalConnection connection) {
+        BsonDocument response = getSpeculativeAuthenticateResponse();
+        if (response != null) {
+            return response;
+        }
+
+        try {
+            byte[] serverResponse = saslClient.hasInitialResponse() ? saslClient.evaluateChallenge(new byte[0]) : null;
+            return sendSaslStart(serverResponse, connection);
+        } catch (Exception e) {
+            throw wrapException(e);
+        }
+    }
+
+    private void getNextSaslResponseAsync(final SaslClient saslClient, final InternalConnection connection,
+                                          final SingleResultCallback<Void> callback) {
+        BsonDocument authenticateResponse = getSpeculativeAuthenticateResponse();
+        try {
+            if (authenticateResponse == null) {
+                byte[] response = (saslClient.hasInitialResponse() ? saslClient.evaluateChallenge(new byte[0]) : null);
+                sendSaslStartAsync(response, connection, new SingleResultCallback<BsonDocument>() {
+                    @Override
+                    public void onResult(final BsonDocument result, final Throwable t) {
+                        if (t != null) {
+                            callback.onResult(null, wrapException(t));
+                        } else if (result.getBoolean("done").getValue()) {
+                            verifySaslClientComplete(saslClient, result, callback);
+                        } else {
+                            new Continuator(saslClient, result, connection, callback).start();
+                        }
+                    }
+                });
+            } else {
+                if (authenticateResponse.getBoolean("done").getValue()) {
+                    verifySaslClientComplete(saslClient, authenticateResponse, callback);
+                } else {
+                    new Continuator(saslClient, authenticateResponse, connection, callback).start();
+                }
+            }
+        } catch (Exception e) {
+            throw wrapException(e);
```

Comment:
Callback wrapped but not passed to the callback. - Does this capture need to be inline into the callbacks?

Reply:
The exception is now passed to the callback.